base/compiler/typeinfer.jl

```diff
@@ -727,15 +727,23 @@ function merge_call_chain!(parent::InferenceState, ancestor::InferenceState, chi
     # and ensure that walking the parent list will get the same result (DAG) from everywhere
     # Also taint the termination effect, because we can no longer guarantee the absence
     # of recursion.
-    tristate_merge!(parent, Effects(EFFECTS_TOTAL, terminates=TRISTATE_UNKNOWN))
+    terminates_globally = is_effect_overrided(parent, :terminates_globally)
+    frames = InferenceState[parent]
     while true
         add_cycle_backedge!(child, parent, parent.currpc)
         union_caller_cycle!(ancestor, child)
-        tristate_merge!(child, Effects(EFFECTS_TOTAL, terminates=TRISTATE_UNKNOWN))
+        terminates_globally |= is_effect_overrided(child, :terminates_globally)
+        push!(frames, child)
         child = parent
         child === ancestor && break
         parent = child.parent::InferenceState
     end
+    if !terminates_globally
+        for frame in frames
+            tristate_merge!(frame, Effects(EFFECTS_TOTAL, terminates=TRISTATE_UNKNOWN))
+        end
+    end
+    return nothing
```

Comment:
With this change, `terminates` is encoded into `ipo_effect_bits` of `:terminates_globally`-functions no matter if they are involved with cycles.
I think it is valid to skip tainting frames in a cycle as long as any of them is annotated as `:terminates_globally`, but I would like to check my understanding with your review, @Keno .

Reply:
you could write malicious functions that violate that, but I don't know if that is the intention of the annotation
```julia
f(x) = while x; g(); end
@assume_effects :terminates_globally g() = f(false)
```

Reply:
Ah, that's a good example, thanks.
So "termination" implied by `:terminates_globally` annotation subsumes constant-prop', but `merge_call_chain!` isn't a good place to handle it. Maybe we just want to just override edge effects in `abstract_call_method` when the edge effect is specified by annotation, rather than trying hard to figure out the appropriate `ipo_purity_bits`s for frames in a cycle.